test/jll.jl

```diff
@@ -0,0 +1,126 @@
+    @test jll_uuid("FFMPEG_jll") == UUID("b22a6f82-2f65-5046-a5b2-351ab43fb4e5")
+
+    project = build_project_dict("LibFoo", v"1.3.5", [Dependency("Zlib_jll"), Dependency(PackageSpec(name = "XZ_jll", version = v"2.4.6"))])
+    @test project["deps"] == Dict("Pkg"      => "44cfe95a-1eb2-52ea-b672-e2afdf69b78f",
+                                  "Zlib_jll" => "83775a58-1f1d-513f-b197-d71354ab007a",
+                                  "Libdl"    => "8f399da3-3557-5675-b5ff-fb832c97cbdb",
+                                  "XZ_jll"   => "ffd25f8a-64ca-5728-b0f7-c24cf3aae800")
+    @test project["name"] == "LibFoo_jll"
+    @test project["uuid"] == "b250f842-3251-58d3-8ee4-9a24ab2bab3f"
+    @test project["compat"] == Dict("julia" => "1.0", "XZ_jll" => "=2.4.6")
+    @test project["version"] == "1.3.5"
+    # Make sure BuildDependency's don't find their way to the project
+    @test_throws MethodError build_project_dict("LibFoo", v"1.3.5", [Dependency("Zlib_jll"), BuildDependency("Xorg_util_macros_jll")])
+end
+
+@testset "JLLs - building" begin
+    mktempdir() do build_path
+        name = "libfoo"
+        version = v"1.0.0"
+        sources = [DirectorySource(build_tests_dir)]
+        # Build for this platform and a platform that isn't this one for sure:
+        # FreeBSD.
+        platforms = [platform, FreeBSD(:x86_64)]
+        dependencies = [Dependency("Zlib_jll")]
+        # The buffer where we'll write the JSON meta data
+        buff = IOBuffer()
+
+        # First pass to get the JSON meta data
+        autobuild(
+            build_path,
+            name,
+            version,
+            sources,
+            # Use the particular build script we're interested in
+            libfoo_make_script,
+            platforms,
+            # The products we expect to be build
+            libfoo_products,
+            dependencies;
+            # Don't do audit passes
+            skip_audit=true,
+            meta_json_stream = buff,
+        )
+
+        # Now build for real
+        build_output_meta = autobuild(
+            build_path,
+            name,
+            version,
+            sources,
+            # Use the particular build script we're interested in
+            libfoo_make_script,
+            platforms,
+            # The products we expect to be build
+            libfoo_products,
+            dependencies;
+            # Don't do audit passes
+            skip_audit=true,
+        )
+
+        withenv(
+            "JULIA_PKG_DEVDIR" => joinpath(build_path, "devdir"),
+            # Let's pretend to be in Yggdrasil, set the relevant environment
+            # variables.
+            "YGGDRASIL" => "true",
+            "BUILD_SOURCEVERSION" => "0123456789abcdef0123456789abcdef01234567",
+            "PROJECT" => "L/$(name)",
+        ) do
+            # What follows loosely mimics what we have to do to build JLL packages
+            # in Yggdrasil.
+            buff = IOBuffer(strip(String(take!(buff))))
```

Comment:
I'm pretty sure that this is overly complicated, but I didn't quickly understand why `buff` as it is caused errors in `JSON.parse` below

Reply:
It looks like the `strip` is mandatory, without it (or simply with `seekstart(buff)` instead of this line) I get
```julia
JLLs - building: Error During Test at /home/mose/.julia/dev/BinaryBuilder/test/jll.jl:23
  Got exception outside of a @test
  Unexpected end of input
   ...when parsing byte with value '10'
  Stacktrace:
   [1] error(::String) at ./error.jl:33
   [2] _error(::String, ::JSON.Parser.StreamingParserState{Base.GenericIOBuffer{Array{UInt8,1}}}) at /home/mose/.julia/packages/JSON/d89fA/src/Parser.jl:150
   [3] byteat(::JSON.Parser.StreamingParserState{Base.GenericIOBuffer{Array{UInt8,1}}}) at /home/mose/.julia/packages/JSON/d89fA/src/Parser.jl:59
   [4] current at /home/mose/.julia/packages/JSON/d89fA/src/Parser.jl:72 [inlined]
   [5] chomp_space! at /home/mose/.julia/packages/JSON/d89fA/src/Parser.jl:117 [inlined]
   [6] parse_value(::JSON.Parser.ParserContext{Dict{String,Any},Int64,true,nothing}, ::JSON.Parser.StreamingParserState{Base.GenericIOBuffer{Array{UInt8,1}}}) at /home/mose/.julia/packages/JSON/d89fA/src/Parser.jl:160
   [7] parse(::Base.GenericIOBuffer{Array{UInt8,1}}; dicttype::Type{T} where T, inttype::Type{Int64}, allownan::Bool, null::Nothing) at /home/mose/.julia/packages/JSON/d89fA/src/Parser.jl:494
   [8] parse(::Base.GenericIOBuffer{Array{UInt8,1}}) at /home/mose/.julia/packages/JSON/d89fA/src/Parser.jl:492
```

Reply:
If you want, you can replace `IOBuffer()` with `BufferStream()` from `SimpleBufferStream` and you probably don't have to do any of this.  Just use `buff` without a `seekstart()` at all.

Reply:
For the record, it turned out `SimpleBufferStream` doesn't help here, it's probably a limitation of `JSON.jl`.  `JSON3.jl` might be a better option, but it also adds some complexity (we'd need to create a struct to wrap the entire data we want to write to disk) that might not be worth the extra work now, since `JSON.jl` isn't perfect but does the hob.